zebra-state/src/service/check/utxo.rs

```diff
@@ -60,52 +107,97 @@ pub fn transparent_double_spends(
-                // to spend its own outputs)
-                if output.tx_index_in_block >= spend_tx_index_in_block {
-                    return Err(EarlyTransparentSpend { outpoint: *spend });
-                } else {
-                    // a unique spend of a previous transaction's output is ok
-                    continue;
-                }
-            }
+            let utxo = transparent_spend_chain_order(
+                *spend,
+                spend_tx_index_in_block,
+                &prepared.new_outputs,
+                non_finalized_chain_unspent_utxos,
+                non_finalized_chain_spent_utxos,
+                finalized_state,
+            )?;
 
-            if non_finalized_chain_spent_utxos.contains(spend) {
-                // reject the spend if its UTXO is already spent in the
-                // non-finalized parent chain
-                return Err(DuplicateTransparentSpend {
-                    outpoint: *spend,
-                    location: "the non-finalized chain",
-                });
-            }
+            // Re-do the coinbase spend restriction checks, as a defence in depth.
+            //
+            // The spendability of valid UTXOs should not change, because
+            // coinbase transaction IDs commit to the block height, and
+            // transaction IDs are unique:
+            // https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0004-asynchronous-script-verification.md#parallel-coinbase-checks
+            //
+            // But the state service returns UTXOs from pending blocks,
+            // which can be rejected by later contextual checks.
+            // This is a particular issue for v5 transactions,
+            // because their authorizing data is only bound to the block data
+            // during contextual validation (#2336).
+            //
+            // To avoid issues like this, we re-check spendability using known valid UTXOs.
 
-            if !non_finalized_chain_unspent_utxos.contains_key(spend)
-                && finalized_state.utxo(spend).is_none()
-            {
-                // we don't keep spent UTXOs in the finalized state,
-                // so all we can say is that it's missing from both
-                // the finalized and non-finalized chains
-                // (it might have been spent in the finalized state,
-                // or it might never have existed in this chain)
-                return Err(MissingTransparentOutput {
-                    outpoint: *spend,
-                    location: "the non-finalized and finalized chain",
-                });
+            let spend_restriction = transaction.coinbase_spend_restriction(prepared.height);
+            if cfg!(not(test)) {
+                // TODO: fix proptests to produce valid UTXO spends (#ticket TODO)
+                //       and unconditionally enable this check
+                validate_transparent_coinbase_spend(*spend, spend_restriction, utxo)?;
```

Comment:
@jvff there are two alternatives here:
1. Just `validate_transparent_coinbase_spend` in the non-finalized state
2. Check it in both the non-finalized state and the UTXO requests 

We can `validate_transparent_coinbase_spend` in the non-finalized state, as long as we fix the proptests to produce valid spends.

If we do the validation here, we get rid of the state request changes in this PR. And we don't have to worry about UTXOs from invalid queued blocks making other invalid blocks pass this check. (Or making valid blocks fail.)

Or we do both checks, which gets rid of most invalid blocks early, and makes sure everything is valid. But it seems to be more complicated than we need.

Fixing the proptests should be doable. We'd need to delete all transparent inputs until we're past 100 blocks, then we substitute valid `OutPoint`s. We'll also need to make the tests use ~110 blocks, so we actually get good coverage. And check what we're doing with proptest genesis outputs.

Reply:
I've updated the PR for option 1 - just validate transparent coinbase spends in the non-finalized state:
* 7a1ba4bc885db78bfdd280c0a44cd6d2d7a0f8e8 is the initial PR split
* 949479577b7f869e6efa3e80a0ec0c9cfdc3a063 does some comment fixes, function renames, and function moves

I still need to:
- [ ] fix the proptest UTXOs
- [ ] make the proptest chain lengths slightly longer (ideally using a single constant)
- [ ] check what the proptests are doing with genesis coinbase outputs